// geofind-contract-generator/src/messageTemplates.js
const nodemailer = require('nodemailer');
require('dotenv').config({ path: __dirname + '/../.env' });

class MessageTemplates {
    constructor() {
        this.senderEmail = process.env.EMAIL_USER;
        this.senderPassword = process.env.EMAIL_PASSWORD;

        this.transporter = nodemailer.createTransport({
            host: "smtp.gmail.com",
            port: 587,
            secure: false,
            auth: { user: this.senderEmail, pass: this.senderPassword }
        });
    }

    async sendAppointmentConfirmation(recipientEmail, clientName, date, time, location) {
        const subject = "✅ Cita Confirmada - Sistema de Citas GeoFind";
        const htmlBody = this.getAppointmentConfirmationHtml(clientName, date, time, location);
        return this._sendEmail(recipientEmail, subject, htmlBody);
    }

    async sendAppointmentRejection(recipientEmail, clientName, reason) {
        const subject = "❌ Cita No Aceptada - Sistema de Citas GeoFind";
        const htmlBody = this.getAppointmentRejectionHtml(clientName, reason);
        return this._sendEmail(recipientEmail, subject, htmlBody);
    }

    getAppointmentConfirmationHtml(clientName, date, time, location) {
        return `
        <!DOCTYPE html>
        <html>
        <head>
            <meta charset="UTF-8">
            <style>
                body {
                    font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
                    background-color: #f0f2f5;
                    padding: 30px;
                    margin: 0;
                }
                .container {
                    max-width: 600px;
                    margin: auto;
                    background: white;
                    padding: 30px;
                    border-radius: 12px;
                    box-shadow: 0 4px 15px rgba(0,0,0,0.1);
                }
                .header {
                    text-align: center;
                    padding-bottom: 20px;
                    border-bottom: 2px solid #e0e0e0;
                }
                .success-title {
                    color: #2e7d32;
                    font-size: 24px;
                    margin: 0;
                }
                .details-box {
                    background: #e8f5e9;
                    padding: 15px;
                    border-radius: 8px;
                    margin-top: 15px;
                    color: #333;
                }
                .cta-button {
                    display: inline-block;
                    background-color: #2e7d32;
                    color: white;
                    padding: 10px 20px;
                    border-radius: 8px;
                    font-weight: bold;
                    font-size: 16px;
                    text-decoration: none;
                }
                .footer {
                    font-size: 12px;
                    color: #999;
                    text-align: center;
                    margin-top: 10px;
                }
            </style>
        </head>
        <body>
            <div class="container">
                <div class="header">
                    <h1 class="success-title">✅ Cita Confirmada</h1>
                    <p style="color:#555; font-size:14px; margin-top:5px;">Sistema de Citas GeoFind</p>
                </div>
                
                <p style="font-size:16px; color:#333; margin-top:20px;">Hola <strong>${clientName}</strong>,</p>
                <p style="font-size:16px; color:#333;">Tu cita ha sido confirmada con los siguientes detalles:</p>
                
                <div class="details-box">
                    <ul style="list-style:none; padding:0; margin:0;">
                        <li style="padding:8px 0; border-bottom:1px solid #c8e6c9;"><strong>📅 Fecha:</strong> ${date}</li>
                        <li style="padding:8px 0; border-bottom:1px solid #c8e6c9;"><strong>⏰ Hora:</strong> ${time}</li>
                        <li style="padding:8px 0;"><strong>📍 Lugar:</strong> ${location}</li>
                    </ul>
                </div>
                
                <p style="margin-top:20px; font-size:15px; color:#555;">
                    <strong>💡 Recomendaciones:</strong><br>
                    • Llega 10 minutos antes de la hora programada<br>
                    • Trae identificación oficial<br>
                    • Prepara tu documentación
                </p>
                
                <div style="text-align:center; margin-top:25px;">
                    <span class="cta-button">¡Te esperamos!</span>
                </div>
                
                <hr style="margin-top:30px; border:0; border-top:1px solid #ddd;">
                <p class="footer">
                    Este correo fue generado automáticamente por el sistema de citas. Por favor no respondas a este mensaje.
                </p>
            </div>
        </body>
        </html>`;
    }

    getAppointmentRejectionHtml(clientName, reason) {
        return `
        <!DOCTYPE html>
        <html>
        <head>
            <meta charset="UTF-8">
            <style>
                body {
                    font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
                    background-color: #fff3f3;
                    padding: 30px;
                    margin: 0;
                }
                .container {
                    max-width: 600px;
                    margin: auto;
                    background: white;
                    padding: 30px;
                    border-radius: 12px;
                    box-shadow: 0 4px 15px rgba(0,0,0,0.1);
                }
                .header {
                    text-align: center;
                    padding-bottom: 20px;
                    border-bottom: 2px solid #f5c6cb;
                }
                .error-title {
                    color: #c62828;
                    font-size: 24px;
                    margin: 0;
                }
                .reason-box {
                    background: #ffebee;
                    padding: 15px;
                    border-radius: 8px;
                    margin-top: 15px;
                    color: #333;
                    border-left: 4px solid #c62828;
                }
                .cta-button {
                    display: inline-block;
                    background-color: #c62828;
                    color: white;
                    padding: 10px 20px;
                    border-radius: 8px;
                    font-weight: bold;
                    font-size: 16px;
                    text-decoration: none;
                }
                .footer {
                    font-size: 12px;
                    color: #999;
                    text-align: center;
                    margin-top: 10px;
                }
            </style>
        </head>
        <body>
            <div class="container">
                <div class="header">
                    <h1 class="error-title">❌ Cita No Aceptada</h1>
                    <p style="color:#555; font-size:14px; margin-top:5px;">Sistema de Citas GeoFind</p>
                </div>
                
                <p style="font-size:16px; color:#333; margin-top:20px;">Hola <strong>${clientName}</strong>,</p>
                <p style="font-size:16px; color:#333;">Lamentamos informarte que tu cita no pudo ser confirmada.</p>
                
                <div class="reason-box">
                    <p style="margin:0; font-size:16px;"><strong>🔍 Razón:</strong> ${reason}</p>
                </div>
                
                <p style="margin-top:20px; font-size:15px; color:#555;">
                    <strong>🔄 Alternativas:</strong><br>
                    • Programa una nueva cita<br>
                    • Contacta para más información<br>
                    • Revisa otros horarios disponibles
                </p>
                
                <div style="text-align:center; margin-top:25px;">
                    <span class="cta-button">Gracias por tu comprensión</span>
                </div>
                
                <hr style="margin-top:30px; border:0; border-top:1px solid #f5c6cb;">
                <p class="footer">
                    Este correo fue generado automáticamente por el sistema de citas. Por favor no respondas a este mensaje.
                </p>
            </div>
        </body>
        </html>`;
    }

    async _sendEmail(to, subject, htmlBody) {
        try {
            const mailOptions = { 
                from: this.senderEmail, 
                to, 
                subject, 
                html: htmlBody 
            };
            await this.transporter.sendMail(mailOptions);
            console.log(`✅ Correo enviado a ${to}`);
            return [true, 'Correo enviado correctamente'];
        } catch (err) {
            console.error(`❌ Error al enviar correo: ${err.message}`);
            return [false, `Error al enviar correo: ${err.message}`];
        }
    }
}

module.exports = MessageTemplates;
